src/Microsoft.IdentityModel.Tokens/InMemoryCryptoProviderCache.cs

```diff
@@ -26,20 +26,57 @@
 //------------------------------------------------------------------------------
 
 using System;
-using System.Collections.Concurrent;
 using System.Globalization;
 using Microsoft.IdentityModel.Logging;
+#if NETSTANDARD2_0 
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+#endif
 
 namespace Microsoft.IdentityModel.Tokens
 {
     /// <summary>
     /// Defines a cache for crypto providers.
     /// Current support is limited to <see cref="SignatureProvider"/> only.
     /// </summary>
-    public class InMemoryCryptoProviderCache : CryptoProviderCache
+    public class InMemoryCryptoProviderCache: CryptoProviderCache, IDisposable
+
     {
-        private ConcurrentDictionary<string, SignatureProvider> _signingSignatureProviders = new ConcurrentDictionary<string, SignatureProvider>();
-        private ConcurrentDictionary<string, SignatureProvider> _verifyingSignatureProviders = new ConcurrentDictionary<string, SignatureProvider>();
+        internal CryptoProviderCacheOptions _cryptoProviderCacheOptions;
+        private bool _disposed = false;
+#if NETSTANDARD2_0 
+        private readonly MemoryCache _signingSignatureProviders;
+        private readonly MemoryCache _verifyingSignatureProviders;
+#elif NET45 || NET461 || NET472
+        private readonly EventBasedLRUCache<string, SignatureProvider> _signingSignatureProviders;
+        private readonly EventBasedLRUCache<string, SignatureProvider> _verifyingSignatureProviders;
+#endif
+
+        /// <summary>
+        /// Creates a new instance of <see cref="InMemoryCryptoProviderCache"/> using the default <see cref="CryptoProviderCacheOptions"/>.
+        /// </summary>
+        public InMemoryCryptoProviderCache() : this(new CryptoProviderCacheOptions())
+        {
+        }
+
+        /// <summary>
+        /// Creates a new instance of <see cref="InMemoryCryptoProviderCache"/> using the specified <paramref name="cryptoProviderCacheOptions"/>.
+        /// </summary>
+        /// <param name="cryptoProviderCacheOptions">The options used to configure the <see cref="InMemoryCryptoProviderCache"/>.</param>
+        public InMemoryCryptoProviderCache(CryptoProviderCacheOptions cryptoProviderCacheOptions)
+        {
+            if (cryptoProviderCacheOptions == null)
+                throw LogHelper.LogArgumentNullException(nameof(cryptoProviderCacheOptions));
+
+            _cryptoProviderCacheOptions = cryptoProviderCacheOptions;
+#if NETSTANDARD2_0
+            _signingSignatureProviders = new MemoryCache(Options.Create(new MemoryCacheOptions() { SizeLimit = _cryptoProviderCacheOptions.SizeLimit }));
+            _verifyingSignatureProviders = new MemoryCache(Options.Create(new MemoryCacheOptions() { SizeLimit = _cryptoProviderCacheOptions.SizeLimit }));
+#elif NET45 || NET461 || NET472
+            _signingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(cryptoProviderCacheOptions.SizeLimit, StringComparer.Ordinal);
```

Comment:
should we add a control that would allow reverting to current behavior? we could put it as a deprecated property from the start.

Reply:
Hmmm... my opinion on this is that it isn't necessary. Even if we put it as a deprecated property from the beginning, removing it would still be considered a breaking change.

If this caching work turns out to be an issue for people using Wilson directly, we can just tell them to revert back to version 6.8.0. If they're using SAL, CryptoProvider caching is turned off as is, so they can simply set that flag back to false in order to revert back to the original behavior.